via/static/scripts/video_player/components/ToastMessages.tsx

```diff
@@ -0,0 +1,148 @@
+import { Callout } from '@hypothesis/frontend-shared';
+import classnames from 'classnames';
+import type { ComponentChildren } from 'preact';
+import { useCallback, useEffect, useRef, useState } from 'preact/hooks';
+
+export type ToastMessage = {
+  id: string;
+  type: 'error' | 'success' | 'notice';
+  message: ComponentChildren;
+
+  /**
+   * Visually hidden messages are announced to screen readers but not visible.
+   * Defaults to false.
+   */
+  visuallyHidden?: boolean;
+
+  /**
+   * Determines if the toast message should be auto-dismissed.
+   * Defaults to true.
+   */
+  autoDismiss?: boolean;
+};
+
+type ToastMessageItemProps = {
+  message: ToastMessage;
+  onDismiss: (id: string) => void;
+};
+
+/**
+ * An individual toast message: a brief and transient success or error message.
+ * The message may be dismissed by clicking on it. `visuallyHidden` toast
+ * messages will not be visible but are still available to screen readers.
+ */
+function ToastMessageItem({ message, onDismiss }: ToastMessageItemProps) {
+  // Capitalize the message type for prepending; Don't prepend a message
+  // type for "notice" messages
+  const prefix =
+    message.type !== 'notice'
+      ? `${message.type.charAt(0).toUpperCase() + message.type.slice(1)}: `
+      : '';
+
+  return (
+    <Callout
+      classes={classnames({
+        'sr-only': message.visuallyHidden,
+      })}
+      status={message.type}
+      onClick={() => onDismiss(message.id)}
+      variant="raised"
+    >
+      <strong>{prefix}</strong>
+      {message.message}
+    </Callout>
+  );
+}
+
+export type ToastMessagesProps = {
+  messages: ToastMessage[];
+  onMessageDismiss: (id: string) => void;
+};
+
+/**
+ * A collection of toast messages. These are rendered within an `aria-live`
+ * region for accessibility with screen readers.
+ */
+export default function ToastMessages({
+  messages,
+  onMessageDismiss,
+}: ToastMessagesProps) {
+  const [dismissedMessages, setDismissedMessages] = useState<string[]>([]);
+  const scheduledMessages = useRef(new Set<string>());
+
+  const dismissMessage = useCallback(
+    (id: string) => {
+      if (dismissedMessages.includes(id)) {
+        return;
+      }
+
+      // Track that this message has been dismissed, and once animation ends,
+      // actually remove it
+      setDismissedMessages(ids => [...ids, id]);
+      setTimeout(() => onMessageDismiss(id), 500);
```

Comment:
IIRC, the entire purpose of a delay on dismiss was to ensure the transition had time to complete before unmounting the message component. I think we have better tools in our toolbox now (e.g. listening for a transition-end event).

What would you think of initially landing a version of `ToastMessages` that doesn't handle dismiss delays or transitions at all, but adds them in a follow-on PR just focused on that? Could we reuse the `TransitionComponent` concept? Perhaps it would be helpful to get the core functionality of these toast messages sorted out without tripping over transitions?

Reply:
Yeah, that sounds reasonable. It would let us land something usable earlier